From f07853917eee092d5dd16feeac4918abd2172128 Mon Sep 17 00:00:00 2001
From: Antony Abee Prakash XV <antonyabee.prakashxv@adlinktech.com>
Date: Fri, 24 Jul 2020 13:27:31 +0530
Subject: [PATCH 14/14] Our SPI is connected differently

- We use SPI1, 2 and 4 instead of just SPI1.
- SPI1 and 2 provide 2 CS lines (i.e. support 2 devices) each.
- The on-module SPI flash is on SPI4, using a single CS. It's a different
  chip, too: Winbond W25Q64 instead of ST M25P32. Apart from the different
  ID this also means the SPI clock can be 3x faster.
- To achieve this speed enable the 'phase-shift SPI clock during data input'
  workaround for 30 MHz and faster clocks.
- A backup SST 25VF032B also exists, and the BMC may switch chips
  automatically. Fortunately Linux will just issue a warning and continue if
  it reads a different chip ID, so there is nothing we have to (or can) do
  about it.
- Very different pinmux, naturally.
- Added tsc2046 Touch screen controller on LEC-Base R1 carrier. Added
  pca9535 GPIO expander in i2c3 bus which will be used for interrupt purpose in
  touch screen controller.
- Added reg_smarc_lcd_vdd needed for touch screen controller.
- Added spidev node for SPI2

Signed-off-by: Antony Abee Prakash XV <antonyabee.prakashxv@adlinktech.com>
---
 arch/arm/boot/dts/lec-imx6.dtsi | 128 +++++++++++++++++++++++---------
 1 file changed, 91 insertions(+), 37 deletions(-)

diff --git a/arch/arm/boot/dts/lec-imx6.dtsi b/arch/arm/boot/dts/lec-imx6.dtsi
index 6dbffc32b318..e010c71197c0 100644
--- a/arch/arm/boot/dts/lec-imx6.dtsi
+++ b/arch/arm/boot/dts/lec-imx6.dtsi
@@ -41,6 +41,15 @@
 		reg = <0x10000000 0x40000000>;
 	};
 
+	reg_smarc_lcd_vdd: regulator-smarc-lcd-vdd {
+		compatible = "regulator-fixed";
+		regulator-name = "smarc_lcd_vdd";
+		gpio = <&gpio1 30 GPIO_ACTIVE_HIGH>; /* SMARC LCD_VDD_EN */
+		enable-active-high;
+		regulator-always-on; /* Freescale's fb/lcd/ldb drivers can't control a regulator yet :-( */
+		regulator-state-mem { regulator-off-in-suspend; }; /* no effect? */
+	};
+
 	reg_smarc_usb0: regulator-smarc-usb0 {
 		compatible = "regulator-fixed";
 		regulator-name = "smarc_usb0";
@@ -386,9 +395,54 @@
 	pinctrl-0 = <&pinctrl_ecspi1>;
 	pinctrl-assert-gpios = <&gpio1 3 GPIO_ACTIVE_LOW>; /* AEN# */
 	status = "okay";
+
+	tsc2046@0 {             /* Touch screen controller on LEC-Base R1 carrier */
+		compatible = "ti,tsc2046";
+		reg = <0>;
+		vcc-supply = <&reg_smarc_lcd_vdd>;
+		spi-max-frequency = <2500000>;
+		interrupt-parent = <&smarc_gpio>;
+		interrupts = <10 IRQ_TYPE_EDGE_FALLING>; /* SMARC GPIO10 */
+		pendown-gpio = <&smarc_gpio 10 GPIO_ACTIVE_LOW>;
+		ti,vref-mv = /bits/ 16 <3300>;
+	};
 };
 #endif
 
+&ecspi2 {	/* SMARC SPI1 */
+	fsl,spi-num-chipselects = <2>;
+	cs-gpios = <&gpio2 27 GPIO_ACTIVE_LOW>, /* SMARC CS0 */
+		<&gpio2 26 GPIO_ACTIVE_LOW>; /* SMARC CS1 */
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_ecspi2>;
+	status = "okay";
+
+	spidev:spidev@0 {
+		status = "okay";
+		compatible = "linux,spidev";
+		spi-max-frequency = <2500000>;
+		reg = <0>;
+	};
+};
+
+&ecspi4 {	/* on-module U-Boot SPI */
+	fsl,spi-num-chipselects = <1>;
+	cs-gpios = <&gpio3 25 GPIO_ACTIVE_LOW>;
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_ecspi4>;
+	fsl,clk-workaround-freq = <30000000>;
+	status = "okay";
+
+	flash: m25p80@0 {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		compatible = "winbond,w25q64";
+		spi-max-frequency = <66000000>; /* slowest we used was SST25VF032B; i.MX6 provides 60MHz max */
+		reg = <0>;
+		m25p,fast-read;
+	};
+};
+
 &fec {
 	pinctrl-names = "default";
 	pinctrl-0 = <&pinctrl_enet>;
@@ -718,41 +772,22 @@
 
 &i2c3 {
 	clock-frequency = <100000>;
-	pinctrl-names = "default", "gpio";
+	pinctrl-names = "default";
 	pinctrl-0 = <&pinctrl_i2c3>;
-	pinctrl-1 = <&pinctrl_i2c3_gpio>;
-	scl-gpios = <&gpio1 3 GPIO_ACTIVE_HIGH>;
-	sda-gpios = <&gpio1 6 GPIO_ACTIVE_HIGH>;
 	status = "okay";
 
-	egalax_ts@4 {
-		compatible = "eeti,egalax_ts";
-		reg = <0x04>;
-		interrupt-parent = <&gpio6>;
-		interrupts = <7 2>;
-		wakeup-gpios = <&gpio6 7 0>;
-	};
-
-	magnetometer@e {
-		compatible = "fsl,mag3110";
-		reg = <0x0e>;
-		pinctrl-names = "default";
-		pinctrl-0 = <&pinctrl_i2c3_mag3110_int>;
-		interrupt-parent = <&gpio3>;
-		interrupts = <16 IRQ_TYPE_EDGE_RISING>;
-		vdd-supply = <&reg_sensors>;
-		vddio-supply = <&reg_sensors>;
-	};
-
-	light-sensor@44 {
-		compatible = "fsl,isl29023";
-		reg = <0x44>;
-		rext = <499>;
-		pinctrl-names = "default";
-		pinctrl-0 = <&pinctrl_i2c3_isl29023_int>;
-		interrupt-parent = <&gpio3>;
-		interrupts = <9 IRQ_TYPE_EDGE_FALLING>;
-		vdd-supply = <&reg_sensors>;
+	smarc_gpio: pca9535@20 {	/* GPIO expander */
+		/* GPIOs 4-6, 8-11 map directly to the SMARC connector's GPIOx.
+		 * GPIOs 0-3, 7 are unused unless a solder jumper is changed, per default
+		 * SMARC GPIO0-3,7 are provided by the i.MX6 (CAM0/1_PWR/RST#,PCAM_FLD). */
+		compatible = "nxp,pca9535";
+		reg = <0x20>;
+		interrupt-parent = <&gpio1>;
+		interrupts = <29 IRQ_TYPE_LEVEL_LOW>; /* GPIO_INT_B */
+		gpio-controller;
+		#gpio-cells = <2>;
+		interrupt-controller;
+		#interrupt-cells = <2>;
 	};
 };
 
@@ -765,9 +800,8 @@
 			fsl,pins = <
 				MX6QDL_PAD_GPIO_0__CCM_CLKO1    0x130b0
 				MX6QDL_PAD_NANDF_CLE__GPIO6_IO07 0x1b0b0
-				MX6QDL_PAD_ENET_TXD1__GPIO1_IO29 0x1b0b0
-				MX6QDL_PAD_EIM_D22__GPIO3_IO22  0x1b0b0
-				MX6QDL_PAD_EIM_CS1__GPIO2_IO24 0x80000000
+				MX6QDL_PAD_ENET_TXD1__GPIO1_IO29 0x80000000 /* GPIO_INT_B */
+				MX6QDL_PAD_ENET_TXD0__GPIO1_IO30 0x80000000 /* SMARC LCD_VDD_EN */
 				MX6QDL_PAD_ENET_RXD0__GPIO1_IO27 0x80000000 /* SMARC USB0_EN_OC# (open-drain) */
 				MX6QDL_PAD_EIM_D29__GPIO3_IO29	0x1a8b0 /* SMARC USB1_EN_OC# (open-drain) */
 				MX6QDL_PAD_EIM_D30__GPIO3_IO30	0x1a8b0 /* SMARC USB2_EN_OC# (open-drain) */
@@ -799,6 +833,26 @@
 			>;
 		};
 #endif
+
+		pinctrl_ecspi2: ecspi2grp {
+			fsl,pins = <	/* SMARC SPI1 */
+				MX6QDL_PAD_EIM_OE__ECSPI2_MISO          0x100b1
+				MX6QDL_PAD_EIM_CS1__ECSPI2_MOSI         0x100b1
+				MX6QDL_PAD_EIM_CS0__ECSPI2_SCLK         0x100b1
+				MX6QDL_PAD_EIM_RW__GPIO2_IO26           0x1b0b0 /* SMARC CS1 */
+				MX6QDL_PAD_EIM_LBA__GPIO2_IO27          0x1b0b0 /* SMARC CS0 */
+			>;
+		};
+
+		pinctrl_ecspi4: ecspi4grp {
+			fsl,pins = <	/* on-module U-Boot SPI */
+				MX6QDL_PAD_EIM_D22__ECSPI4_MISO         0x100b1
+				MX6QDL_PAD_EIM_D28__ECSPI4_MOSI         0x100b1
+				MX6QDL_PAD_EIM_D21__ECSPI4_SCLK         0x100b1
+				MX6QDL_PAD_EIM_D25__GPIO3_IO25          0x1b0b0 /* CS3 */
+			>;
+		};
+
 		pinctrl_enet: enetgrp {
 			fsl,pins = <
 				MX6QDL_PAD_ENET_MDIO__ENET_MDIO		0x1b0b0
@@ -901,8 +955,8 @@
 
 		pinctrl_i2c3: i2c3grp {
 			fsl,pins = <
-				MX6QDL_PAD_GPIO_3__I2C3_SCL		0x4001b8b1
-				MX6QDL_PAD_GPIO_6__I2C3_SDA		0x4001b8b1
+				MX6QDL_PAD_EIM_D17__I2C3_SCL		0x4001b8b1
+				MX6QDL_PAD_EIM_D18__I2C3_SDA		0x4001b8b1
 			>;
 		};
 
-- 
2.17.1

